File: simba-zookeeper/src/test/java/me/ahoo/simba/zookeeper/ZookeeperMutexContendServiceTest.java
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.simba.zookeeper;

import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import me.ahoo.simba.core.AbstractMutexContender;
import me.ahoo.simba.core.MutexState;

import org.apache.curator.framework.CuratorFramework;
import org.apache.curator.framework.CuratorFrameworkFactory;
import org.apache.curator.retry.RetryNTimes;
import org.apache.curator.test.TestingServer;
import org.junit.jupiter.api.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.jupiter.api.Assertions.*;

/**
 * @author ahoo wang
 */
@Slf4j
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ZookeeperMutexContendServiceTest {
    
    private CuratorFramework curatorFramework;
    private ZookeeperMutexContendServiceFactory contendServiceFactory;
    TestingServer testingServer;
    
    @SneakyThrows
    @BeforeAll
    void setup() {
        testingServer = new TestingServer();
        testingServer.start();
        curatorFramework = CuratorFrameworkFactory.newClient(testingServer.getConnectString(), new RetryNTimes(1, 10));
        curatorFramework.start();
        this.contendServiceFactory = new ZookeeperMutexContendServiceFactory(curatorFramework);
    }
    
    @SneakyThrows
    @AfterAll
    void after() {
        if (Objects.nonNull(curatorFramework)) {
            curatorFramework.close();
        }
        if (Objects.nonNull(testingServer)) {
            testingServer.stop();
        }
    }
    
    @Test
    void start() {
        String mutex = "start";
        
        CompletableFuture<MutexState> acquiredFuture = new CompletableFuture();
        CompletableFuture<MutexState> releasedFuture = new CompletableFuture();
        
        ZookeeperMutexContendService contendService = (ZookeeperMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
            @Override
            public void onAcquired(MutexState mutexState) {
                log.info("onAcquired");
                acquiredFuture.complete(mutexState);
            }
            
            @Override
            public void onReleased(MutexState mutexState) {
                log.info("onReleased");
                releasedFuture.complete(mutexState);
            }
        });
        
        contendService.start();
        acquiredFuture.join();
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture.join();
        assertFalse(contendService.isOwner());
    }
    
    
    @Test
    void restart() {
        String mutex = "restart";
        CompletableFuture<MutexState> acquiredFuture = new CompletableFuture();
        CompletableFuture<MutexState> releasedFuture = new CompletableFuture();
        
        CompletableFuture<MutexState> acquiredFuture2 = new CompletableFuture();
        CompletableFuture<MutexState> releasedFuture2 = new CompletableFuture();
        ZookeeperMutexContendService contendService = (ZookeeperMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
            @Override
            public void onAcquired(MutexState mutexState) {
                log.info("onAcquired");
                if (!acquiredFuture.isDone()) {
                    acquiredFuture.complete(mutexState);
                } else {
                    acquiredFuture2.complete(mutexState);
                }
            }
            
            @Override
            public void onReleased(MutexState mutexState) {
                log.info("onReleased");
                if (!releasedFuture.isDone()) {
                    releasedFuture.complete(mutexState);
                } else {
                    releasedFuture2.complete(mutexState);
                }
            }
        });
        
        contendService.start();
        acquiredFuture.join();
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture.join();
        assertFalse(contendService.isOwner());
        
        contendService.start();
        acquiredFuture2.join();
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture2.join();
        assertFalse(contendService.isOwner());
    }
    
    @SneakyThrows
    @Test
    void guard() {
        String mutex = "guard";
        CompletableFuture<MutexState> acquiredFuture = new CompletableFuture();
        CompletableFuture<MutexState> releasedFuture = new CompletableFuture();
        
        ZookeeperMutexContendService contendService = (ZookeeperMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
            @Override
            public void onAcquired(MutexState mutexState) {
                log.info("onAcquired");
                acquiredFuture.complete(mutexState);
            }
            
            @Override
            public void onReleased(MutexState mutexState) {
                log.info("onReleased");
                releasedFuture.complete(mutexState);
            }
        });
        
        contendService.start();
        assertTrue(acquiredFuture.join().isOwner(contendService.getContenderId()));
        TimeUnit.SECONDS.sleep(3);
        assertEquals(contendService.getAfterOwner().getOwnerId(), contendService.getContender().getContenderId());
        assertTrue(contendService.isOwner());
        contendService.stop();
        releasedFuture.join();
        assertFalse(contendService.isOwner());
    }
    
    @Test
    void multiContend() throws InterruptedException {
        String mutex = "multiContend";
        AtomicInteger count = new AtomicInteger(0);
        AtomicReference<String> currentOwnerIdRef = new AtomicReference<>();
        List<ZookeeperMutexContendService> contendServiceList = new ArrayList<>(10);
        for (int i = 0; i < 10; i++) {
            ZookeeperMutexContendService contendService = (ZookeeperMutexContendService) contendServiceFactory.createMutexContendService(new AbstractMutexContender(mutex) {
                @Override
                public void onAcquired(MutexState mutexState) {
                    currentOwnerIdRef.set(mutexState.getAfter().getOwnerId());
                    super.onAcquired(mutexState);
                    assertEquals(1, count.incrementAndGet());
                }
                
                @Override
                public void onReleased(MutexState mutexState) {
                    super.onReleased(mutexState);
                    assertEquals(0, count.decrementAndGet());
                }
            });
            contendService.start();
            contendServiceList.add(contendService);
        }
        TimeUnit.SECONDS.sleep(30);
        assertEquals(1, count.get());
        String currentOwnerId = currentOwnerIdRef.get();
        for (ZookeeperMutexContendService contendService : contendServiceList) {
            if (contendService.hasOwner()) {
                Assertions.assertEquals(contendService.getAfterOwner().getOwnerId(), currentOwnerId);
            }
        }
        assertEquals(1, contendServiceList.stream().filter(contendService -> contendService.getContenderId().equals(currentOwnerId)).count());
    }
    
}

File: simba-zookeeper/src/test/kotlin/me/ahoo/simba/zookeeper/ZookeeperMutexContendServiceTest.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.simba.zookeeper

import me.ahoo.simba.core.MutexContendServiceFactory
import me.ahoo.simba.core.MutexContender
import me.ahoo.simba.core.MutexOwner
import me.ahoo.simba.core.MutexRetrievalService.Status
import me.ahoo.simba.core.MutexState
import me.ahoo.simba.test.MutexContendServiceSpec
import org.apache.curator.framework.CuratorFramework
import org.apache.curator.framework.CuratorFrameworkFactory
import org.apache.curator.retry.RetryNTimes
import org.apache.curator.test.TestingServer
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import java.util.concurrent.Executor
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.atomic.AtomicInteger

/**
 * Thread-safe counting contender: LeaderLatch callbacks arrive on the Curator
 * EventThread while assertions poll from the test thread.
 */
private class CountingContender(
    override val mutex: String,
    override val contenderId: String
) : MutexContender {
    val acquiredCount = AtomicInteger()

    override fun onAcquired(mutexState: MutexState) {
        acquiredCount.incrementAndGet()
    }

    override fun onReleased(mutexState: MutexState) = Unit
}

/**
 * @author ahoo wang
 */

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
internal class ZookeeperMutexContendServiceTest : MutexContendServiceSpec() {
    lateinit var curatorFramework: CuratorFramework
    override lateinit var mutexContendServiceFactory: MutexContendServiceFactory
    lateinit var testingServer: TestingServer

    @BeforeAll
    fun setup() {
        testingServer = TestingServer()
        testingServer.start()
        curatorFramework = CuratorFrameworkFactory.newClient(testingServer.connectString, RetryNTimes(1, 10))
        curatorFramework.start()
        mutexContendServiceFactory = ZookeeperMutexContendServiceFactory(ForkJoinPool.commonPool(), curatorFramework)
    }

    @AfterAll
    fun destroy() {
        if (this::curatorFramework.isInitialized) {
            curatorFramework.close()
        }
        if (this::testingServer.isInitialized) {
            testingServer.stop()
        }
    }

    @Test
    fun `late isLeader callback after stop must not revive ownership`() {
        val mutex = "late-is-leader-${System.currentTimeMillis()}"
        val contender = CountingContender(mutex, "late-callback-contender")
        // synchronous handleExecutor: notifyOwner applies inline on the calling thread
        val directExecutor = Executor { it.run() }
        val contendService = ZookeeperMutexContendService(contender, directExecutor, curatorFramework)

        contendService.start()

        // the sole contender genuinely becomes leader shortly after start
        val deadline = System.currentTimeMillis() + 10_000
        while (contender.acquiredCount.get() == 0 && System.currentTimeMillis() < deadline) {
            Thread.sleep(50)
        }
        assertThat("sole contender should become leader", contender.acquiredCount.get(), equalTo(1))

        contendService.stop()
        assertThat(contendService.status, equalTo(Status.INITIAL))
        assertThat("state after stop should be owner-less", contendService.afterOwner, equalTo(MutexOwner.NONE))

        // simulate a LeaderLatch callback dispatched by the Curator EventThread
        // racing with close() and arriving after stop completed
        contendService.isLeader()

        assertThat(
            "late isLeader must be ignored after stop",
            contender.acquiredCount.get(),
            equalTo(1)
        )
        assertThat("mutexState must stay owner-less after stop", contendService.afterOwner, equalTo(MutexOwner.NONE))
    }
}
